### PR TITLE
Change HDF5Type by parsing `base` into object fields

### PR DESCRIPTION
**Pull request by loichuder**
A refactoring in preparation of #406.

`HDF5FloatType` and `HDF5IntegerType` no longer holds a `base` string but directly the information that can be deduced from it (endianness, number of bits, sign). This will allow a much easier handling in the `JupyterProvider` (for type inferring) and in the `MetadataViewer` (for displaying the HDF5Type of a dataset).

The `base` string is now parsed in the `HsdsProvider` to add the relevant fields to `HDF5Type`. Note that some advanced types that hold reference to other types need to be recursively parsed.

### CODE REVIEW

File: src/h5web/providers/hdf5-models.ts [hunk 1]
```diff
@@ -103,40 +103,20 @@ export enum HDF5TypeClass {
...
-    | 'H5T_IEEE_F32LE'
-    | 'H5T_IEEE_F64BE'
-    | 'H5T_IEEE_F64LE';
+  length: number;
+  endianness: Endianness;
+}
+
+export enum Endianness {
+  Big = 'Big',
+  Little = 'Little',
```
Comment (loichuder) [inline src/h5web/providers/hdf5-models.ts:119]: `enum` is perhaps too much here...

File: src/h5web/providers/hsds/models.ts [hunk 2]
```diff
@@ -1,15 +1,49 @@
...
+  HDF5TypeClass,
+  HDF5StringType,
+  HDF5Dims,
 } from '../hdf5-models';
 import type { Dataset, Group } from '../models';
 
+export type HsdsBaseType =
+  | { class: HDF5TypeClass.Integer; base: string }
+  | { class: HDF5TypeClass.Float; base: string }
+  | HDF5StringType;
```
Comment (loichuder) [inline src/h5web/providers/hsds/models.ts:18]: I left `HDF5StringType` unchanged for now.

File: src/h5web/providers/hsds/utils.ts [hunk 3]
```diff
@@ -26,3 +35,55 @@ export function assertHsdsDataset(
...
+  }
+
+  const [, sign, length, endian] = matches;
+
+  return {
+    class: hsdsClass,
+    endianness: endian === 'L' ? Endianness.Little : Endianness.Big,
+    length: Number.parseInt(length, 10),
+    unsigned: sign === 'U' || undefined,
+  };
```
Comment (loichuder) [inline src/h5web/providers/hsds/utils.ts:60]: I intend to add some tests for this once we wrap up this PR

File: src/h5web/providers/hsds/utils.ts [hunk 4]
```diff
@@ -26,3 +35,55 @@ export function assertHsdsDataset(
...
+}
+
+export function processHsdsType(hsdsType: HsdsType): HDF5Type {
+  if (isString(hsdsType)) {
+    return hsdsType;
```
Comment (loichuder) [inline src/h5web/providers/hsds/utils.ts:65]: If `hsdsType` is a ref to another type (`HDF5Id`)

File: src/h5web/providers/hsds/utils.ts [hunk 5]
```diff
@@ -26,3 +35,55 @@ export function assertHsdsDataset(
...
+    hsdsType.class === HDF5TypeClass.Array ||
+    hsdsType.class === HDF5TypeClass.VLen
+  ) {
+    return {
+      ...hsdsType,
+      base: processHsdsBaseType(hsdsType.base),
+    };
+  }
+
+  if (hsdsType.class === HDF5TypeClass.Compound) {
```
Comment (loichuder) [inline src/h5web/providers/hsds/utils.ts:78]: For compound types, the `type` needs to be processed for each field

File: src/h5web/providers/mock/metadata-utils.ts [hunk 6]
```diff
@@ -29,12 +30,14 @@ import { mockValues } from './values';
...
-  base: 'H5T_STD_I32LE',
+  endianness: Endianness.Little,
+  length: 32,
 };
 
 export const floatType: HDF5FloatType = {
   class: HDF5TypeClass.Float,
-  base: 'H5T_IEEE_F64LE',
+  endianness: Endianness.Little,
+  length: 64,
```
Comment (loichuder) [inline src/h5web/providers/mock/metadata-utils.ts:40]: This improves a little the readability of mocked types

File: src/h5web/providers/hsds/utils.ts [hunk 7]
```diff
@@ -26,3 +35,55 @@ export function assertHsdsDataset(
...
 }
+
+export function processHsdsBaseType(hsdsBaseType: HsdsBaseType): HDF5BaseType {
+  if (hsdsBaseType.class === HDF5TypeClass.String) {
+    return hsdsBaseType;
+  }
+
+  const { class: hsdsClass, base } = hsdsBaseType;
+
+  const regex = /H5T_[A-z]+_([A-z])(\d+)(B|L)E/u;
```
Comment (axelboc) [inline src/h5web/providers/hsds/utils.ts:46]: - I don't think this works with all the types listed here: https://support.hdfgroup.org/HDF5/doc/RM/PredefDTypes.html -- it works with `H5T_STD` and `H5T_IEEE` for sure, so may as well be specific about it.
- The base is all uppercase, so the character ranges can be shortened.
- It might be simpler to just have `(BE|LE)` at the end and just store that directly on the `endianness` prop, since we don't care much about it. 🤷 

So in summary: `/H5T_(STD|IEEE)_([A-Z])(\d+)(BE|LE)/u`

File: src/h5web/providers/hsds/utils.ts [hunk 8]
```diff
@@ -26,3 +35,55 @@ export function assertHsdsDataset(
...
+    return {
+      ...hsdsType,
+      fields: hsdsType.fields.map((v) => ({
+        name: v.name,
+        type: processHsdsType(v.type),
+      })),
+    };
+  }
+
+  return processHsdsBaseType(hsdsType);
```
Comment (axelboc) [inline src/h5web/providers/hsds/utils.ts:88]: Wouldn't this cause infinite recursion?

File: src/h5web/providers/hsds/utils.ts [hunk 9]
```diff
@@ -26,3 +35,55 @@ export function assertHsdsDataset(
...
+
+  if (hsdsType.class === HDF5TypeClass.Compound) {
+    return {
+      ...hsdsType,
+      fields: hsdsType.fields.map((v) => ({
+        name: v.name,
+        type: processHsdsType(v.type),
+      })),
+    };
+  }
```
Comment (axelboc) [inline src/h5web/providers/hsds/utils.ts:86]: These `if`s may be better written as a `switch`. Also, with `...hsdsType`, you keep the `base` property in the returned objects, which is unnecessary -- you could destructure `hsdsType` to get the `class` and specify it explicitly in each returned object.

**Review by axelboc — APPROVED** [review]

File: src/h5web/providers/hsds/utils.ts (hunk 8 above)
Comment (loichuder) [reply src/h5web/providers/hsds/utils.ts]: Nay: `processHsdsType` (now `convertHsdsType`) calls `processHsdsBaseType`

File: src/h5web/providers/hdf5-models.ts [hunk 10]
```diff
@@ -101,42 +101,19 @@ export enum HDF5TypeClass {
   Compound = 'H5T_COMPOUND',
 }
 
+export type Endianness = 'BE' | 'LE';
```
Comment (axelboc) [inline src/h5web/providers/hdf5-models.ts:104]: `HDF5Endianness` ?

File: src/h5web/providers/hsds/utils.ts [hunk 11]
```diff
@@ -26,3 +30,66 @@ export function assertHsdsDataset(
...
+
+  const regex = /H5T_(?:STD|IEEE)_([A-Z])(\d+)(BE|LE)/u;
+  const matches = regex.exec(base);
+
+  if (!matches) {
+    throw new Error(`Unrecognized base ${base}`);
+  }
+
+  const [, sign, size, endianness] = matches;
+  assertEndianness(endianness);
```
Comment (axelboc) [inline src/h5web/providers/hsds/utils.ts:49]: It's a case where it makes sense to do `endianness: endianness as HDF5Endianness`, because it cannot be anything else (unless there were to be a bug in the regex, in which case we'd fix it)

File: src/h5web/providers/hsds/utils.ts [hunk 12]
```diff
@@ -26,3 +30,66 @@ export function assertHsdsDataset(
...
+        class: HDF5TypeClass.Array,
+        base: convertHsdsBaseType(hsdsType.base),
+        dims: hsdsType.dims,
+      };
+
+    case HDF5TypeClass.VLen:
+      return {
+        class: HDF5TypeClass.VLen,
+        base: convertHsdsBaseType(hsdsType.base),
+      };
```
Comment (axelboc) [inline src/h5web/providers/hsds/utils.ts:76]: You can have two `case` in a row:

```
case HDF5TypeClass.Array:
case HDF5TypeClass.VLen:
  return { class: hsdsType.class, ... }
```

**Review by axelboc — APPROVED** [review]

File: src/h5web/providers/hsds/utils.ts (hunk 12 above)
Comment (loichuder) [reply src/h5web/providers/hsds/utils.ts:76]: There is a `dims` field for `HDF5TypeClass.Array` and not for `HDF5TypeClass.VLen`, right ?

Comment (axelboc) [reply src/h5web/providers/hsds/utils.ts:76]: yeah, but you had them both in the same `if` earlier, didn't you?

Comment (axelboc) [reply src/h5web/providers/hsds/utils.ts:76]: nevermind if it's clearer like this, though

Comment (loichuder) [reply src/h5web/providers/hsds/utils.ts:76]: > yeah, but you had them both in the same `if` earlier, didn't you?

Yes because I was doing `...hsdsType` but I also find it much clearer like this.
